src/App.jsx

```diff
@@ -3,25 +3,37 @@ import "./App.css";
 import AddTodoForm from "./AddTodoForm";
 import TodoList from "./TodoList";
 
-const useSemiPersistentState = () => {
-  const [todoList, setTodoList] = useState(() => {
-    try {
-      return JSON.parse(localStorage.getItem("savedTodoList")) || [];
-    } catch (e) {
-      console.error("Failed to parse 'savedTodoList'", e);
-      return [];
-    }
-  });
+const savedTodoList = "savedTodoList";
 
-  useEffect(() => {
-    localStorage.setItem("savedTodoList", JSON.stringify(todoList));
-  }, [todoList]);
+function App() {
+  const [todoList, setTodoList] = useState([]);
+  const [isLoading, setIsLoading] = useState(true);
 
-  return [todoList, setTodoList];
-};
+  useEffect(() => {
+    new Promise((resolve) => {
+      setTimeout(() => {
+        resolve({
+          data: {
+            todoList: () => {
+              try {
+                return JSON.parse(localStorage.getItem(savedTodoList)) || [];
+              } catch (e) {
+                console.error(`Failed to parse ${savedTodoList}`, e);
+                return [];
+              }
+            },
+          },
+        });
+      }, 2000);
+    }).then((result) => {
+      setTodoList(result.data.todoList);
+      setIsLoading(false);
+    });
+  }, []);
 
-function App() {
-  const [todoList, setTodoList] = useSemiPersistentState();
+  useEffect(() => {
+    !isLoading && localStorage.setItem(savedTodoList, JSON.stringify(todoList));
+  }, [todoList]);
```

Comment:
I think you also want to add ``isLoading`` in the dependency array, since you are using isLoading in useEffect.